queries/haskell/highlights.scm

```diff
@@ -190,26 +176,44 @@
          ]))
      (quasiquote)
      ((exp_name) . (operator))
-    ]))
+    ])+
+ (#eq? @variable @_name))
+
+((signature name: (variable) @variable)
+ . (comment)*
+ . (function 
+    name: (variable) @_name
+    rhs: (exp_infix 
+      [
+       (exp_literal)
+       (exp_apply 
+         (exp_name 
+           [(constructor)
+            (variable)
+           ]))
+       (quasiquote)
+       ((exp_name) . (operator))
+      ]))+
  (#eq? @variable @_name))
 
 ((signature name: (variable) @function)
- (function 
-  name: (variable) @_name
-  patterns: (patterns))
+ . (comment)*
+ . (function 
+    name: (variable) @_name
+    patterns: (patterns))+
  (#eq? @function @_name))
 
-(signature 
-  name: (variable) @function
-  type: (fun))
-((signature 
-   name: (variable) @_name
-   type: (fun))
- (function
-   name: (variable) @function)
+;; For some reason, `(signature name: (variable) type: (fun))` doesn't work here
+(signature (variable) @function . (fun))
+
+((signature (variable) @_name . (fun))
+ . ((comment)*
+ . (function (variable)))*
+ . (function (variable) @function)
```

Comment:
This last one doesn't seem necessary since there's already a `(function name: (variable) @function)` right below for all patterns of that

Reply:
Yup. They were just missing ` . (comment)* `.

Reply:
@lucario387 do you happen to know why the following signature queries (~line 241) are anchored to functions?

```scheme
((signature (variable) @function (fun)) . (function (variable)))
((signature (variable) @function (context (fun))) . (function (variable)))
((signature (variable) @function (forall (context (fun)))) . (function (variable)))
```

I believe this schould suffice:

```scheme
((signature (variable) @function (fun)))
((signature (variable) @function (context (fun))))
((signature (variable) @function (forall (context (fun)))))
```

I suppose it was done because they are each accompanied by something like this

```scheme
((signature (variable) @_type (fun)) . (function (variable) @function) (#eq? @function @_type))
```

Personally, I think a signature that has a function type, such as

```haskell
someFunc :: Int -> ReturnType
```

should always be highlighted as a function, even if it's accompanied by a function that for some reason doesn't get highlighted correctly.

The reason I'm asking is because I think I can get rid of the signature queries with long lists of rhs (see [your other comment](https://github.com/nvim-treesitter/nvim-treesitter/pull/5502#discussion_r1351612064)) by adding something like this:

```scheme
;; Consider signatures with only one value on the rhs as variables
(signature . (variable) @variable . (_) . )
;; but consider 'IO a' a function
(signature name: (variable) @function 
  . (type_apply (type_name) @_type)
    (#eq? @_type "IO"))
```

Which would result in

```haskell
intVal :: Int
-- ^ @variable

mbInt :: Maybe Int
-- ^ @variable

intFunc :: Int -> Int
-- ^ @function

getArgs :: IO [String]
-- ^ @function

main :: IO ()
-- ^ @function
```

... but this only works if I remove the anchors in the other queries.

The other function queries seem pretty reliable by now (apart from very few edge cases, so I don't think we really need to check for accompanying signatures).

I'll be dogfeeding the idea this week, but I wanted to get your input on it :smile:

Reply:
You are the maintainer for haskell queries now. If anything, you have the call to change some of the queries 😄 

For the anchored signature, I honestly do not know. If you believe the query can be improved, go for it 😄 . I'm all rooting for you